### PR TITLE
Add "group_by" and "remove" to functional

### PR DESCRIPTION
**Pull request by jackamo**
This PR adds four functions to `functional` and `tablex`:

In `tablex` it adds:
* `get` - equivalent to `tbl[x]` but allows you to set a context-relevant default.
* `get-in` which is similar, but is for nested tables:

```lua
 local get_example = { a = 1, b = 2, c = 3}
print(tablex.get(get_example, "a")) -- => "1"
print(tablex.get(get_example, "b")) -- => "2"
print(tablex.get(get_example, "d")) -- => "nil"
print(tablex.get(get_example, "d", "NOT_FOUND")) -- => "NOT_FOUND"

local get_in_example = {a = {b = {c = {d = "DONE"}}}}
print(tablex.get_in(get_in_example, {"x"}, "NOT_FOUND")) -- => "NOT_FOUND"
print(tablex.get_in(get_in_example, {"a", "b", "d" }, "NOT_FOUND")) -- => "NOT_FOUND"
print(tablex.get_in(get_in_example, {"a", "b", "c", "d"}, "NOT_FOUND")) -- => "DONE"
print(tablex.get_in(get_in_example, {"x"})) -- => "nil"
````

In `functional` it adds:

* `remove` which is just the complement to `filter`
* `group-by` which will return a table keyed by the return of the provided function:

```lua
local functional_example = {{alive = false}, {alive = false}, {alive = false}, {alive = false}, {alive = true}{alive = true}}

functional.group_by(functional_example, function(m) return m.alive end)

-- returns:
--- {true = {{alive = true}, {alive = true}}
---  false = {{alive = false}, {alive = false}, {alive = false}, {alive = false}}}

functional.remove(functional_example, function(m) return not m.alive end)

-- returns:
---- {{alive = true}, {alive = true}}
```

I was a little bored this morning, so I thought I'd add some of the functions I enjoy from my time with Clojure. I'm unsure if they fit with your idea of what `batteries` should be, so happy to be told to go add these somewhere else :)

(Also, sorry for the empty email, I accidentally mashed enter before ready)

### CODE REVIEW

**Comment by jackamo** [comment]
I'm Zombrodo / JackARRR#5614 on discord if you wanna chat / ask questions or anything!

**Comment by 1bardesign** [comment]
No worries about the early email :+1: 

First: I'm happy to have contributions! Thanks for your interest.

A lot of this fits the idea of what `batteries` should be; both `remove` and `group_by` are straightforward, nice functional additions!

However, I have a few performance and correctness concerns :sweat_smile: 

`remove` would be better implemented as its own loop as each call generates a closure with the current implementation, on top of the table generated by `filter`. You'll notice that the library generally takes this approach across implementations, which makes the library code less self-referential and faster at runtime for the user.

`get_in` doesn't need a separate __recurse implementation; the depth can be safely defaulted from `nil` when it's not passed, tail calls are mandatorily optimised in lua, and the "private" function just makes the code harder to read.

`get` and `get_in` taking a default seems harmless until you consider the implications of passing a table initialiser (`{}`) or an anonymous function; that parameter is created anew for every base level invocation. For example, in the case of the `group_by` implementation, this generates garbage every loop iteration.

The more idiomatic `t[k] or {}` doesn't create a table if `t[k]` already exists and is truthy, but still has the correctness issue.
The correctness issue is that `t[k] or d` incorrectly short circuits for `false` values of `t[k]`, not just `nil` values. If you want to support `false` values in the table you need to compare against `nil` directly and this cannot be expressed in a single `and`/`or` ternary style expression in lua afaik; it needs a separate conditional.

For library code I would prefer a more performant implementation for `group_by` and to probably not include `get`, as it encourages hidden garbage generation.

Here's a benchmark to demonstrate the practical difference; in one the default has to be prepared ahead of time. The garbage collector is run outside of the loop to give each test a fresh workspace. It's not exactly rigorous but it demonstrates the problem.
```lua
local runs = 1e5
local elems = 1e3
local missing_elems = 0.5

local working_set = {}
local all_elems = {}
for i = 1, elems do
	local k = tostring(i)
	table.insert(all_elems, k)
	if love.math.random() < missing_elems then
		working_set[k] = {i}
	end
end

function passed_to_function(t, e)
	local f = function(t, k, d)
		local v = t[k]
		if v == nil then
			v = d
		end
		return v
	end
	local sum = 0
	for i, k in ipairs(e) do
		sum = sum + #f(t, k, {})
	end
	return sum
end

function inline(t, e)
	local sum = 0
	for i, k in ipairs(e) do
		local v = t[k]
		if v == nil then
			v = {}
		end
		sum = sum + #v
	end
	return sum
end

for i, v in ipairs({
	{"passed", passed_to_function},
	{"inline", inline},
}) do
	local name, f = unpack(v)
	collectgarbage("collect")
	local start = love.timer.getTime()
	for _ = 1, runs do
		f(working_set, all_elems)
	end
	print(
		("%-20s: %07fms"):format(
			name,
			(love.timer.getTime() - start) * 1000
		)
	)
end

love.event.quit()
```
For me when run with `love` this gives:
```
passed              : 8673.635801ms
inline              : 3560.121445ms
```
in effect, the inline default is more than twice as fast for a table default. If you change to a numeric or boolean default you'll see a much smaller difference; strings are somewhere in between when testing for me but are still much faster than tables. This shows that the issue is in garbage generation, not the overhead of a function call.

tl;dr:
- I'd love to merge this and contributions are welcome!
- to avoid excess garbage generation, please:
  - revise `remove` as a raw loop
  - remove `get` as it encourages the anti-pattern of creating default arguments even when they are usually not used
  - revise `get_in` as a single recursive or iterative function
  - revise `group_by` without `get`

I'm aware this is a non-trivial revision so happy to make the changes myself if you're ok with that, it'll happen at some point this week if so.

If the PR is updated before then, that's awesome and I'll get it merged asap :wink: 

If you'd prefer some other alternative, let me know.

Sorry for the long reply! :sparkling_heart:

**Comment by 1bardesign** [comment]
As a stray thought, `get_in` could probably be implemented in a nice recursive manner with a signature along the lines of `get_in(t, a, ...)`, ie stripping out one argument from the varargs each recursion.

This will be fast because of tail-call optimisation and should be pretty clean!

**Comment by jackamo** [comment]
Great! Awesome feedback.

Definitely a lot of the functionality is pulled from Clojure, and that has a lot of in-built optimisations (such as first-class immutability, and persistent/transient maps) which makes a lot of this efficient there, but I have found in the past that wrangling Lua is a quite a bit different.

I wanted to implement a `first` and `rest` function using unpack, but then went down the NYI rabbit hole, and relied on the recursive approach for `get-in`. 

But overall you've cleared up some of the fears regarding performance, thanks for spending time on 'em! I'll revise the code, and get back to you.

**Comment by jackamo** [comment]
>  implications of passing a table initialiser ({}) or an anonymous function;

It's so easy to forget that anonymous functions / closures aren't nice in Lua, and that tables are everything. I was ready to discuss that I didn't see it as an anti pattern, but no, you're right! There's defintely more of a performance cost there than I was thinking. Other than returning a number, boolean or string literal, it's pretty non-performant.

**Comment by 1bardesign** [comment]
👍 Sounds good! NYI is indeed a rabbit hole and easy to run afoul of, sometimes without noticing cause the interpreter is still more than fast enough for smaller workloads.

`batteries` code doesn't have to be the fastest at all costs*, I'm just hesitant to bake too many performance pitfalls in, haha.

* I generally prefer ipairs for readability over a numeric for on tables for instance

**Comment by jackamo** [comment]
> As a stray thought, get_in could probably be implemented in a nice recursive manner

So I've implemented both an iterative and recursive solution, but I'm trying to work out which would be more useful - one that takes varargs (which is elegant) or one that takes a table of keys, which allows keys to be programatically created, and passed.

```
function tablex.get_in(t, k, ...)
  if not ... then
    return t[k]
  end
  return tablex.get_in(t[k], ...)
end
```

The issue is that if you wanted to programatically use `get_in` with this use case, you'd have to call `unpack` and we know that's no fun. While I could fudge it with a type check, I don't like it when a function can take two distinct types (despite being a dynamically typed language). 

I guess if I can't answer it with a handy use case and certainty, it perhaps shouldn't be in the PR :) I might actually remove them, and just stick with `group_by` and `remove`.

**Comment by 1bardesign** [comment]
I'd say that switching on the type of the argument k and iterating if it's a table would be no big deal. I can see the appeal of both; varargs strings seems most performant and simplest for inline use but a table is definitely better for building up something complex with code. I haven't used code like either much (maybe it'd be good for pulling data out of json payloads or similar?) so I'm not sure which is more often used in practice!

Could be worth moving to a separate issue/pr to get the rest merged while you think about it though!

**Comment by jackamo** [comment]
Yeah - I've removed the commit, and the remainder has been amended, please feel free to give feedback back on those when you can :)

File: functional.lua [hunk 1]
```diff
@@ -263,4 +274,16 @@ function functional.find_match(t, f)
...
+-- returns a table where the elements in t are grouped by the result of f on each element.
+function functional.group_by(t, f)
+	local result = {}
+	for i, v in ipairs(t) do
+		local group = f(v)
+		if result[group] == nil then
+			result[group] = {}
+		end
+		table.insert(result[group], v)
+	end
```
Comment (1bardesign) [inline functional.lua:286]: Probably needs to `return result` here 😉

Comment (jackamo) [reply functional.lua:286]: Curse you Lisp for having me forget how to write `return` 😞

**Comment by 1bardesign** [comment]
I've had to make a few changes, they'll be pushed shortly

- renamed `remove` to `remove_if` because `batteries:export()` overlays `functional` onto `table`, and `table.remove` already exists. If you have a better name a PR is welcome :wink: 
- moved `group_by` to right below partition, as it seems it belongs there more than at the end of the file
- added a little more to the documentation

In any case, lets get this merged :+1: :sparkling_heart: :sparkles:

**Comment by jackamo** [comment]
Wicked, thanks again for the feedback! I agree with all your changes :)
